### PR TITLE
table partition: update job dependence for exchange partition with table

### LINKED ISSUE #35842: table partition: update job dependence for exchange partition with table
## Enhancement

exchange partition with table involves two tables, so the current job dependence function needs to be modified.

### PR DESCRIPTION
**Pull request by ymkzpx**
### What problem does this PR solve?

Issue Number: close #35842

### What is changed and how it works?

update job dependence for exchange partition with table.

### Check List

Tests 

None.

Side effects

None.

Documentation

None.

### Release note

```release-note
None.
```

### CODE REVIEW

File: parser/model/model_test.go [hunk 1]
```diff
@@ -274,6 +274,101 @@ func TestJobCodec(t *testing.T) {
 	require.NoError(t, err)
 	require.True(t, isDependent)
 
+	// Test IsDependentOn for exchange partition with table.
+	// test ActionCreateSchema and ActionExchangeTablePartition is dependent.
```
Comment (mjonss) [inline parser/model/model_test.go:278]: I think ActionDropSchema and ActionDropTable are also interesting to test for completeness.

**Review by mjonss — APPROVED** [review]
LGTM, just some smaller suggestions/questions.

File: parser/model/ddl.go [hunk 2]
```diff
@@ -572,6 +572,63 @@ func (job *Job) hasDependentSchema(other *Job) (bool, error) {
 				return true, nil
 			}
 		}
+		if job.Type == ActionExchangeTablePartition {
+			var (
+				defID          int64
+				ptSchemaID     int64
+				ptID           int64
+				partName       string
+				withValidation bool
```
[the remark this answers is not in the document]
Comment (ymkzpx) [reply parser/model/ddl.go:581]: Exchange partition DDL job has five args. If only three parameters are passed during the `DecodeArgs`, the remaining two parameters are discarded. As a result, only three parameters can be obtained during the `DecodeArgs` elsewhere.

File: parser/model/ddl.go [hunk 3]
```diff
@@ -572,6 +572,63 @@ func (job *Job) hasDependentSchema(other *Job) (bool, error) {
...
+			)
+			if err := job.DecodeArgs(&defID, &ptSchemaID, &ptID, &partName, &withValidation); err != nil {
```
[the remark this answers is not in the document]
Comment (ymkzpx) [reply parser/model/ddl.go:583]: same above

File: parser/model/ddl.go [hunk 4]
```diff
@@ -572,6 +572,63 @@ func (job *Job) hasDependentSchema(other *Job) (bool, error) {
...
+}
+
+func (job *Job) hasDependentTableForExchangePartition(other *Job) (bool, error) {
+	if job.Type == ActionExchangeTablePartition {
+		var (
+			defID          int64
+			ptSchemaID     int64
+			ptID           int64
+			partName       string
+			withValidation bool
```
[the remark this answers is not in the document]
Comment (ymkzpx) [reply parser/model/ddl.go:601]: Same as above.

File: parser/model/model_test.go (hunk 1 above)
Comment (ymkzpx) [reply parser/model/model_test.go:278]: thanks, it has been added.

File: parser/model/ddl.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mjonss) [reply parser/model/ddl.go:583]: Ack, thanks for the explanation.

File: parser/model/ddl.go [hunk 5]
```diff
@@ -578,6 +578,84 @@ func (job *Job) hasDependentSchema(other *Job) (bool, error) {
...
+			}
+			if defID == other.TableID || defID == otherPtID || defID == otherDefID {
+				return true, nil
+			}
+		}
+	}
+	return false, nil
+}
+
+func (job *Job) hasDependentSchemaForExchangePartition(other *Job) (bool, error) {
```
Comment (xhebox) [inline parser/model/ddl.go:642]: The function is unused? Or you forgot to replace the above code with this method?

File: parser/model/ddl.go [hunk 6]
```diff
@@ -578,6 +578,84 @@ func (job *Job) hasDependentSchema(other *Job) (bool, error) {
...
+			if other.SchemaID == ptSchemaID {
+				return true, nil
+			}
+		}
+	}
+	return false, nil
+}
+
+func (job *Job) hasDependentTableForExchangePartition(other *Job) (bool, error) {
+	if job.Type == ActionExchangeTablePartition {
```
Comment (xhebox) [inline parser/model/ddl.go:601]: If the current job is not `exchange partition`, will it be blocked by the other `exchange partition` job?

File: parser/model/ddl.go (hunk 5 above)
Comment (ymkzpx) [reply parser/model/ddl.go]: yes, has been deleted.

File: parser/model/ddl.go (hunk 6 above)
Comment (ymkzpx) [reply parser/model/ddl.go:606]: Thank you for reminding me. As follows, I modified it in this way.
```go
isDependent, err = job.hasDependentTableForExchangePartition(other)
if err != nil || isDependent {
    return isDependent, errors.Trace(err)
}
isDependent, err = other.hasDependentTableForExchangePartition(job)
if err != nil || isDependent {
    return isDependent, errors.Trace(err)
}
```

Comment (xhebox) [reply parser/model/ddl.go:606]: Add a test case for that?

Comment (ymkzpx) [reply parser/model/ddl.go:606]: done.

**Review by xhebox — APPROVED** [review]

**Pull request merged**